File: BlockTutorial/BlockTutorial/HTTPBinManagerOperation.m
//
//  RecipetUploadOperation.m
//  BlockTutorial
//
//  Created by w91379137 on 2016/6/27.
//  Copyright © 2016年 w91379137. All rights reserved.
//

#import "PDSSetting.h"
#import "HTTPBinManagerOperation.h"
#import "WebService.h"

@interface HTTPBinManagerOperation()

@end

@implementation HTTPBinManagerOperation

#pragma mark -
- (void)main
{
    @autoreleasepool {
        self.error = nil;
        [self loopAllTask];
    }
}

#pragma mark - Task
- (void)loopAllTask
{
    if (self.isCancelled || self.error) return;
    
    weakSelfMake(weakSelf);
    if (!self.getDict) [self taskGet:^{ [weakSelf loopAllTask]; }];
    else if (!self.postDict) [self taskPost:^{ [weakSelf loopAllTask]; }];
    else if (!self.image) [self taskImage:^{ [weakSelf loopAllTask]; }];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.delegate operationNotice:self];
    });
}

- (void)taskGet:(void(^)())callback
{
    [WebService fetchGetResponseWithCallback:^(NSDictionary *dict, NSError *err) {
        if (err) {
            self.error  = err;
            [self cancel];
        }
        else {
            self.getDict = dict;
        }
        
        if(callback) callback();
    }];
}

- (void)taskPost:(void(^)())callback
{
    [WebService postCustomerName:@"test"
                        callback:^(NSDictionary *dict, NSError *err) {
                            
                            if (err) {
                                self.error  = err;
                                [self cancel];
                            }
                            else {
                                self.postDict = dict;
                            }
                            
                            if(callback) callback();
                        }];
}

- (void)taskImage:(void(^)())callback
{
    [WebService fetchImageWithCallback:^(UIImage *image, NSError *err) {
        
        if (err) {
            self.error  = err;
            [self cancel];
        }
        else {
            self.image = image;
        }
        
        if(callback) callback();
    }];
}

#pragma mark -
- (void)cancel
{
    [super cancel]; //cancel self no error
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.delegate operationNotice:self];
    });
}

- (float)progress
{
    float progress = 0;
    if (self.getDict) progress += .33;
    if (self.postDict) progress += .33;
    if (self.image) progress += .34;
    return progress;
}

@end

File: GCDProgramming30/GCDProgramming30/HTTPBinManagerOperation.swift
//
//  HTTPBinManagerOperation.swift
//  GCDProgramming23
//
//  Created by w91379137 on 2016/6/28.
//  Copyright © 2016年 w91379137. All rights reserved.
//

import UIKit

protocol HTTPBinManagerOperationDelegate {
    func operationNotice<T : HTTPBinManagerOperationType>(operation: T!)
}

protocol HTTPBinManagerOperationType {
    
    var isCancelled: Bool { get }
    func progress() -> Float
    var error : NSError? { get }
}

class HTTPBinManagerOperation: Operation,HTTPBinManagerOperationType {
    
    //MARK: - Property
    var delegate : HTTPBinManagerOperationDelegate?
    
    var error : NSError?
    var image : UIImage?
    var getDict : NSDictionary?
    var postDict : NSDictionary?
    
    //MARK:
    override func main() {
        self.loopAllTask()
    }
    
    func loopAllTask() {
        
        if self.isCancelled || (self.error != nil) { return }
        DispatchQueue.main.async() {
            self.delegate?.operationNotice(operation: self);
        }
        
        if self.getDict == nil {
            fetchGetResponseWithCallback { (dict, error) in
                if error != nil {
                    self.error = error
                    self.cancel()
                }
                else {
                    self.getDict = dict
                }
                self.loopAllTask()
            }
        }
        else if self.postDict == nil {
            postCustomerName(name: "test") { (dict, error) in
                if error != nil {
                    self.error = error
                    self.cancel()
                }
                else {
                    self.postDict = dict
                }
                self.loopAllTask()
            }
        }
        else if self.image == nil {
            fetchImageWithCallback { (image, error) in
                if error != nil {
                    self.error = error
                    self.cancel()
                }
                else {
                    self.image = image
                }
                self.loopAllTask()
            }
        }
        else {
            //Finish
        }
    }
    
    //MARK:
    override func cancel() {
        super.cancel()
        self.delegate?.operationNotice(operation: self)
    }
    
    func progress() -> Float {
        var progress = Float(0)
        if self.getDict != nil {progress += 0.33}
        if self.postDict != nil {progress += 0.33}
        if self.image != nil {progress += 0.34}
        
        return progress
    }
}


